### PR TITLE
[BEAM-9703]Include user distritribution into metric-dedicated validate runner test.

### PR DESCRIPTION
**Pull request by HuangLED**
Include user distribution into PY validation runner test dedicated for metrics. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by HuangLED** [comment]
@lukecwik

**Comment by lukecwik** [comment]
retest this please

File: sdks/python/apache_beam/metrics/metric_test.py [hunk 1]
```diff
@@ -162,6 +169,17 @@ def process(self, element):
...
+    # Verify user distribution counter.
+    metric_results = res.metrics().query()
+    namespace = 'apache_beam.metrics.metric_test.SomeDoFn'
+    hc.assert_that(
+        metric_results['distributions'],
+        hc.contains_inanyorder(
+            MetricResult(
+                MetricKey('ApplyPardo', MetricName(namespace, 'element_dist')),
+                DistributionResult(DistributionData(10, 4, 1, 4)),
+                DistributionResult(DistributionData(10, 4, 1, 4)))))
```
Comment (lukecwik) [inline sdks/python/apache_beam/metrics/metric_test.py:181]: not all runners support attempted metrics, please only compare against committed metrics for now

Comment (HuangLED) [reply sdks/python/apache_beam/metrics/metric_test.py:181]: I see. Thanks for the information! 

I checked MetricResult interface, it only works when passed in both 'commited' and 'attempted'. 

I will take a look what alternative we need to do.  If you by any chance are aware of a particular API I should use, please let me know.

Comment (lukecwik) [reply sdks/python/apache_beam/metrics/metric_test.py:181]: Write your own matcher.

Comment (HuangLED) [reply sdks/python/apache_beam/metrics/metric_test.py]: Sure. 

Done update. 

Please take a look if it makes sense.

File: sdks/python/apache_beam/metrics/execution.py [hunk 2]
```diff
@@ -106,11 +107,13 @@ def __init__(self, key, committed, attempted):
...
     self.committed = committed
     self.attempted = attempted
+    self.committed_only = committed_only
 
   def __eq__(self, other):
     return (
         self.key == other.key and self.committed == other.committed and
-        self.attempted == other.attempted)
+        (self.committed_only or other.committed_only or
+         self.attempted == other.attempted))
```
Comment (lukecwik) [inline sdks/python/apache_beam/metrics/execution.py:116]: This is not what I meant when I suggested a matcher, see https://github.com/hamcrest/PyHamcrest#writing-custom-matchers for an example.

Comment (HuangLED) [reply sdks/python/apache_beam/metrics/execution.py]: By searching for MatcherBase, found an existing implementation under testing folder. 

Updated.   PTAL.

**Comment by lukecwik** [comment]
retest this please

**Comment by lukecwik** [comment]
retest this please

**Comment by HuangLED** [comment]
lint and test both work fine on my local repository. 

And the PythonLint by github reports from a different file:
00:42:57 apache_beam/transforms/environments.py:255:12: F821 undefined name 'from_container_image'
00:42:57     return from_container_image(
00:42:57            ^
00:42:57 1     F821 undefined name 'from_container_image'

Not sure why.  Looking.

**Comment by HuangLED** [comment]
pull from head and rebased (no changes to previous commits)

please try again

**Comment by lukecwik** [comment]
retest this please

**Comment by lukecwik** [comment]
> gradlew lint and test both work fine on my local repository.
> 
> test command: python setup.py nosetests --tests apache_beam.metrics.metric_test:MetricsTest.test_user_counter_using_pardo
> 
> And the PythonLint by github, is from a different file(not touched by this PR):
> 00:42:57 apache_beam/transforms/environments.py:255:12: F821 undefined name 'from_container_image'
> 00:42:57 return from_container_image(
> 00:42:57 ^
> 00:42:57 1 F821 undefined name 'from_container_image'
> 
> Not sure why. Looking.

This is a known issue. Note you never need to pull from head unless there is a conflict since your changes will be automatically applied on to what is currently at HEAD when the tests are run.

**Comment by HuangLED** [comment]
> > gradlew lint and test both work fine on my local repository.
> > test command: python setup.py nosetests --tests apache_beam.metrics.metric_test:MetricsTest.test_user_counter_using_pardo
> > And the PythonLint by github, is from a different file(not touched by this PR):
> > 00:42:57 apache_beam/transforms/environments.py:255:12: F821 undefined name 'from_container_image'
> > 00:42:57 return from_container_image(
> > 00:42:57 ^
> > 00:42:57 1 F821 undefined name 'from_container_image'
> > Not sure why. Looking.
> 
> This is a known issue. Note you never need to pull from head unless there is a conflict since your changes will be automatically applied on to what is currently at HEAD when the tests are run.

Good to know. Thanks!

**Comment by lukecwik** [comment]
Run PythonLint PreCommit

**Comment by lukecwik** [comment]
Run Python PreCommit

**Comment by lukecwik** [comment]
Run Portable_Python PreCommit

**Comment by lukecwik** [comment]
Run PythonLint PreCommit

**Comment by lukecwik** [comment]
Run Portable_Python PreCommit

**Comment by lukecwik** [comment]
Run Python PreCommit

**Comment by lukecwik** [comment]
Run Python PreCommit

**Comment by lukecwik** [comment]
Run Portable_Python PreCommit

**Comment by lukecwik** [comment]
Run PythonLint PreCommit

**Pull request merged**